File: tutorial/T28-Stage.cpp
/***************************
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T28-Stage.cpp
@Time: 2024/12/15 18:40
@Desc: 本例主要展示在多个并发的element中，通过 GStage 做同步的功能
***************************/

#include "MyGNode/MyNode1.h"
#include "MyGNode/MyStageNode.h"

using namespace CGraph;

void tutorial_stage() {
    GPipelinePtr pipeline = GPipelineFactory::create();
    GElementPtr a, b, c, d, e = nullptr;

    pipeline->registerGElement<MyNode1>(&a, {}, "nodeA");
    pipeline->registerGElement<MyStageNode<1, 2>>(&b, {a}, "nodeB");
    pipeline->registerGElement<MyStageNode<3, 1>>(&c, {a}, "nodeC");
    pipeline->registerGElement<MyStageNode<3, 1>>(&d, {a}, "nodeD");
    pipeline->registerGElement<MyNode1>(&e, {b, c, d}, "nodeE");

    /**
     * 添加一个 stage 信息，遇到 3次的时候，会通知触发结束
     * 正好对应上面三个 MyStageNode 的逻辑
     */
    pipeline->addGStage<GStage>(kStageKey, 3);

    pipeline->process();
    GPipelineFactory::remove(pipeline);
}


int main() {
    tutorial_stage();
    return 0;
}

File: python/tutorial/T28-Stage.py
"""
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T28-Stage
@Time: 2025/4/4 01:08
@Desc: 
"""

from pycgraph import GPipeline, GStage

from MyGNode.MyNode1 import MyNode1
from MyGNode.MyStageNode import MyStageNode


def tutorial_stage():
    pipeline = GPipeline()
    a, b, c, d, e = MyNode1(), MyStageNode(1, 2), MyStageNode(3, 1), MyStageNode(3, 1), MyNode1()

    pipeline.registerGElement(a, set(), "nodeA")
    pipeline.registerGElement(b, {a}, "nodeB")
    pipeline.registerGElement(c, {a}, "nodeC")
    pipeline.registerGElement(d, {a}, "nodeD")
    pipeline.registerGElement(e, {b, c, d}, "nodeE")

    stage = GStage()
    pipeline.addGStage(stage, 'stage1', 3)

    pipeline.process()


if __name__ == '__main__':
    tutorial_stage()
